File: fastapi_app/examples/test_client.ts
/**
 * TypeScript Test Client for FastAPI Structural Design API.
 *
 * Demonstrates how to use:
 * - REST endpoints for beam design
 * - WebSocket for interactive design
 * - SSE for batch design streaming
 *
 * Usage:
 *   # With Node.js (requires 'ws' package):
 *   npx ts-node test_client.ts
 *
 *   # Or compile and run:
 *   tsc test_client.ts && node test_client.js
 *
 * Week 3 Implementation - V3 Migration
 */

// Types
interface BeamDesignRequest {
  width: number;
  depth: number;
  moment: number;
  shear?: number;
  fck: number;
  fy: number;
}

interface FlexureResult {
  ast_required: number;
  ast_provided: number;
  is_safe: boolean;
  utilization_ratio: number;
}

interface DesignResult {
  status: 'PASS' | 'FAIL';
  flexure: FlexureResult;
  shear?: object;
}

interface HealthResponse {
  status: string;
  version: string;
  timestamp: string;
}

// Configuration
const BASE_URL = process.env.API_URL || 'http://localhost:8000';
const WS_URL = BASE_URL.replace('http', 'ws');

// =============================================================================
// REST API Client
// =============================================================================

export class StructuralDesignClient {
  private baseUrl: string;

  constructor(baseUrl: string = BASE_URL) {
    this.baseUrl = baseUrl;
  }

  /**
   * Check API health status.
   */
  async health(): Promise<HealthResponse> {
    const response = await fetch(`${this.baseUrl}/health`);
    if (!response.ok) {
      throw new Error(`Health check failed: ${response.status}`);
    }
    return response.json();
  }

  /**
   * Design a reinforced concrete beam.
   */
  async designBeam(params: BeamDesignRequest): Promise<DesignResult> {
    const response = await fetch(`${this.baseUrl}/api/v1/design/beam`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(params),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Design failed: ${error.detail || response.status}`);
    }

    return response.json();
  }

  /**
   * Get beam geometry calculations.
   */
  async calculateGeometry(
    width: number,
    depth: number,
    length: number,
  ): Promise<{
    volume: number;
    surface_area: number;
    weight: number;
  }> {
    const response = await fetch(
      `${this.baseUrl}/api/v1/geometry/beam?width=${width}&depth=${depth}&length=${length}`,
    );

    if (!response.ok) {
      throw new Error(`Geometry calculation failed: ${response.status}`);
    }

    return response.json();
  }
}

// =============================================================================
// WebSocket Client
// =============================================================================

export class LiveDesignClient {
  private ws: WebSocket | null = null;
  private baseUrl: string;

  constructor(baseUrl: string = WS_URL) {
    this.baseUrl = baseUrl;
  }

  /**
   * Connect to WebSocket for live design.
   */
  async connect(clientId: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const url = `${this.baseUrl}/ws/design/${clientId}`;
      this.ws = new WebSocket(url);

      this.ws.onopen = () => {
        console.log(`✅ Connected to ${url}`);
        resolve();
      };

      this.ws.onerror = (error) => {
        reject(new Error(`WebSocket error: ${error}`));
      };
    });
  }

  /**
   * Send a design request and wait for response.
   */
  async designBeam(params: BeamDesignRequest): Promise<DesignResult> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Design request timeout'));
      }, 5000);

      this.ws!.onmessage = (event) => {
        clearTimeout(timeout);
        const data = JSON.parse(event.data);
        if (data.type === 'design_result') {
          resolve(data.data);
        } else if (data.type === 'error') {
          reject(new Error(data.message));
        }
      };

      this.ws!.send(
        JSON.stringify({
          type: 'design_beam',
          params,
        }),
      );
    });
  }

  /**
   * Close the WebSocket connection.
   */
  close(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

// =============================================================================
// SSE Batch Client
// =============================================================================

export class BatchDesignClient {
  private baseUrl: string;

  constructor(baseUrl: string = BASE_URL) {
    this.baseUrl = baseUrl;
  }

  /**
   * Stream batch design results via SSE.
   */
  async *streamBatchDesign(
    beams: Array<BeamDesignRequest & { id: string }>,
  ): AsyncGenerator<{ event: string; data: object }> {
    const beamsJson = encodeURIComponent(JSON.stringify(beams));
    const url = `${this.baseUrl}/stream/batch-design?beams=${beamsJson}`;

    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Batch design failed: ${response.status}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error('No response body');
    }

    const decoder = new TextDecoder();
    let buffer = '';
    let currentEvent = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('event:')) {
          currentEvent = line.slice(6).trim();
        } else if (line.startsWith('data:')) {
          const data = JSON.parse(line.slice(5).trim());
          yield { event: currentEvent, data };
          currentEvent = '';
        }
      }
    }
  }
}

// =============================================================================
// Demo
// =============================================================================

async function main() {
  console.log('=' .repeat(60));
  console.log('🧪 TypeScript FastAPI Client Demo');
  console.log('=' .repeat(60));
  console.log(`\nServer: ${BASE_URL}`);

  const client = new StructuralDesignClient();

  // Test health
  console.log('\n📤 GET /health');
  try {
    const health = await client.health();
    console.log(`📥 Status: ${health.status}`);
  } catch (e) {
    console.log(`❌ ${e}`);
    console.log('   Make sure the server is running first!');
    return;
  }

  // Test beam design
  console.log('\n📤 POST /api/v1/design/beam');
  try {
    const result = await client.designBeam({
      width: 300,
      depth: 500,
      moment: 150,
      fck: 25,
      fy: 500,
    });
    console.log(`📥 Status: ${result.status}`);
    console.log(`   Ast required: ${result.flexure.ast_required.toFixed(1)} mm²`);
  } catch (e) {
    console.log(`❌ ${e}`);
  }

  console.log('\n✅ Demo complete!');
}

// Run if executed directly (Node.js)
if (typeof require !== 'undefined' && require.main === module) {
  main().catch(console.error);
}

File: fastapi_app/examples/test_client.py
#!/usr/bin/env python3
"""
Python Test Client for FastAPI Live Design Endpoints.

Demonstrates how to use:
- WebSocket for interactive design (bi-directional)
- SSE for batch design (server → client stream)

Usage:
    # Start the server first:
    uvicorn fastapi_app.main:app --reload --host 0.0.0.0 --port 8000

    # Then run this client:
    python fastapi_app/examples/test_client.py

Week 3 Implementation - V3 Migration
"""

from __future__ import annotations

import asyncio
import json
import sys
from typing import Any
from urllib.parse import quote

import httpx

# Default server URL
BASE_URL = "http://localhost:8000"
WS_URL = "ws://localhost:8000"


# =============================================================================
# WebSocket Client
# =============================================================================


async def test_websocket_design():
    """
    Test WebSocket live design endpoint.

    Demonstrates:
    - Connection lifecycle
    - Sending design requests
    - Receiving instant results
    """
    try:
        import websockets
    except ImportError:
        print("❌ websockets package not installed. Run: pip install websockets")
        return

    print("\n" + "=" * 60)
    print("🔌 Testing WebSocket Live Design")
    print("=" * 60)

    uri = f"{WS_URL}/ws/design/test-client-1"

    try:
        async with websockets.connect(uri) as ws:
            print(f"✅ Connected to {uri}")

            # Test 1: Ping
            print("\n📤 Sending ping...")
            await ws.send(json.dumps({"type": "ping"}))
            response = await ws.recv()
            data = json.loads(response)
            print(f"📥 Received: {data['type']} at {data['timestamp']}")

            # Test 2: Design beam
            beam_params = {
                "width": 300,
                "depth": 500,
                "moment": 150,
                "shear": 75,
                "fck": 25,
                "fy": 500
            }
            print(f"\n📤 Designing beam: {beam_params}")
            await ws.send(json.dumps({
                "type": "design_beam",
                "params": beam_params
            }))
            response = await ws.recv()
            data = json.loads(response)

            print(f"📥 Result in {data['latency_ms']:.1f}ms:")
            print(f"   Ast required: {data['data']['flexure']['ast_required']:.1f} mm²")
            print(f"   Safe: {data['data']['flexure']['is_safe']}")

            # Test 3: Another design with higher moment
            beam_params["moment"] = 250
            print(f"\n📤 Designing beam with higher moment: {beam_params['moment']} kN·m")
            await ws.send(json.dumps({
                "type": "design_beam",
                "params": beam_params
            }))
            response = await ws.recv()
            data = json.loads(response)

            print(f"📥 Result in {data['latency_ms']:.1f}ms:")
            print(f"   Ast required: {data['data']['flexure']['ast_required']:.1f} mm²")

            print("\n✅ WebSocket test complete!")

    except ConnectionRefusedError:
        print(f"❌ Could not connect to {uri}")
        print("   Make sure the server is running: uvicorn fastapi_app.main:app --reload")


# =============================================================================
# SSE Client
# =============================================================================


async def test_sse_batch_design():
    """
    Test SSE batch design endpoint.

    Demonstrates:
    - Streaming multiple beam designs
    - Progress tracking
    - Event parsing
    """
    print("\n" + "=" * 60)
    print("📡 Testing SSE Batch Design")
    print("=" * 60)

    beams = [
        {"id": "B1", "width": 300, "depth": 500, "moment": 100, "fck": 25, "fy": 500},
        {"id": "B2", "width": 350, "depth": 550, "moment": 150, "fck": 25, "fy": 500},
        {"id": "B3", "width": 400, "depth": 600, "moment": 200, "fck": 30, "fy": 500},
        {"id": "B4", "width": 300, "depth": 500, "moment": 120, "fck": 25, "fy": 500},
        {"id": "B5", "width": 350, "depth": 600, "moment": 180, "fck": 30, "fy": 500},
    ]

    beams_json = json.dumps(beams)
    url = f"{BASE_URL}/stream/batch-design?beams={quote(beams_json)}"

    print(f"📤 Designing {len(beams)} beams via SSE...")
    print()

    try:
        async with httpx.AsyncClient() as client:
            async with client.stream("GET", url) as response:
                current_event = None

                async for line in response.aiter_lines():
                    if line.startswith("event:"):
                        current_event = line.split(":")[1].strip()
                    elif line.startswith("data:"):
                        data = json.loads(line[5:].strip())

                        if current_event == "start":
                            print(f"🚀 Job started: {data['job_id']} ({data['total']} beams)")

                        elif current_event == "design_result":
                            status = "✅" if data["status"] == "PASS" else "❌"
                            print(f"   {status} {data['beam_id']}: Ast={data['flexure']['ast_required']:.0f} mm²")

                        elif current_event == "progress":
                            pct = data["percent"]
                            bar = "█" * int(pct / 5) + "░" * (20 - int(pct / 5))
                            print(f"   [{bar}] {pct:.0f}%", end="\r")

                        elif current_event == "complete":
                            print()
                            print(f"\n✅ Batch complete!")
                            print(f"   Total: {data['completed']}/{data['total']}")
                            print(f"   Failed: {data['failed']}")
                            print(f"   Duration: {data['duration_seconds']:.2f}s")

                        elif current_event == "error":
                            print(f"   ❌ Error: {data.get('message', 'Unknown error')}")

                        current_event = None

        print("\n✅ SSE test complete!")

    except httpx.ConnectError:
        print(f"❌ Could not connect to {BASE_URL}")
        print("   Make sure the server is running: uvicorn fastapi_app.main:app --reload")


# =============================================================================
# REST API Test
# =============================================================================


async def test_rest_endpoints():
    """Test basic REST endpoints."""
    print("\n" + "=" * 60)
    print("🌐 Testing REST Endpoints")
    print("=" * 60)

    async with httpx.AsyncClient() as client:
        try:
            # Health check
            print("\n📤 GET /health")
            response = await client.get(f"{BASE_URL}/health")
            print(f"📥 {response.status_code}: {response.json()['status']}")

            # Design beam
            print("\n📤 POST /api/v1/design/beam")
            response = await client.post(
                f"{BASE_URL}/api/v1/design/beam",
                json={"width": 300, "depth": 500, "moment": 150, "fck": 25, "fy": 500}
            )
            if response.status_code == 200:
                data = response.json()
                print(f"📥 {response.status_code}: Ast={data['flexure']['ast_required']:.1f} mm²")
            else:
                print(f"📥 {response.status_code}: {response.json()}")

            print("\n✅ REST test complete!")

        except httpx.ConnectError:
            print(f"❌ Could not connect to {BASE_URL}")
            print("   Make sure the server is running: uvicorn fastapi_app.main:app --reload")


# =============================================================================
# Main
# =============================================================================


async def main():
    """Run all tests."""
    print("=" * 60)
    print("🧪 FastAPI Live Design Test Client")
    print("=" * 60)
    print(f"\nServer: {BASE_URL}")
    print("Make sure the server is running before testing!")

    # Test REST
    await test_rest_endpoints()

    # Test SSE
    await test_sse_batch_design()

    # Test WebSocket (requires websockets package)
    await test_websocket_design()

    print("\n" + "=" * 60)
    print("🎉 All tests complete!")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())
